Add docs about how to resize volumes in GSP (example: Concourse Workers)

issyl0 opened the pull request:
- This was a long and manual process, with lots of waiting for pipelines
  to run. Here are some words to tell people what to do next time a disk
  fills up.

issyl0 committed:
Add docs about how to resize volumes in GSP (example: Concourse Workers)

- This was a long and manual process, with lots of waiting for pipelines
  to run. Here are some words to tell people what to do next time a disk
  fills up.

Co-Authored-By: Daniel Blair <blairboy362@users.noreply.github.com>

blairboy362 commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
````diff
@@ -0,0 +1,98 @@
+      $ gds sandbox kubectl -n gsp-system edit persistentvolumeclaim concourse-work-dir-gsp-concourse-worker-2
+      ```
+
+      This will open `$EDITOR`. Find the line with the old volume size,
````
```suggestion
      This will open `$EDITOR`. Update `.spec.resources.requests.storage` to the desired size, then save.
```

blairboy362 commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
````diff
@@ -0,0 +1,98 @@
+      ```
+
+      This will open `$EDITOR`. Find the line with the old volume size,
+      and change it to the desired size, then save.
````
```suggestion
```

issyl0 commented:
I squashed Daniel's edits into the original commit to make for a cleaner history - thanks Daniel!

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,96 @@
+# Resizing persistent volumes
+
+The volumes you created in the beginning may not be big enough for
```
In the beginning of what? When you first create the cluster? Or something else?

blairboy362 replied:
When the thing using the persistent volume was first created...not sure how best to phrase that. Suggestions welcome!

jonathanglassman replied:
I would rephrase to "You may need to resize your persistent volumes as your needs change."

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,96 @@
+your ongoing needs.
+
+As an example, we're going to double the space on the Concourse Worker
+persistent volumes - from 64Gi to 128Gi. Replace
```
If gigabytes, use GB

blairboy362 replied:
Can we compromise on [GiB](https://www.iec.ch/si/binary.htm)?

jonathanglassman replied:
Ok

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
````diff
@@ -0,0 +1,96 @@
+    size: 128Gi
+```
+
+1. When merged, the PR will be released to sandbox. Things must pass
````
Which things? Can we be more specific?

blairboy362 replied:
Changed to:
"Changes must pass through sandbox via the full release process..."

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,96 @@
+
+   The `-deployer` pipeline will re-apply the statefulset. This will
+   intentionally roll all the associated pods.
+
```
Content best practice is to put the action as close to the start of the numbered step as possible. The context is good, I would recommend rearranging the other content. For example:

"Pods in stateful sets each have a volume. For Kubernetes to be able to resize the volumes, we need to detach those volumes from the pods by destroying the stateful set. To destroy the stateful set, run:

    ```sh
    $ gds sandbox kubectl -n gsp-system delete statefulset --cascade=false gsp-concourse-worker
    ```
You should destroy the stateful set when the Concourse release pipeline starts to fail. You will see an error message similar to:

    > The StatefulSet “gsp-concourse-worker” is invalid: spec:
    > Forbidden: updates to statefulset spec for fields other than
    > ‘replicas’, ‘template’, and ‘updateStrategy’ are forbidden

The `--cascade=false` flag makes sure that you do not delete the pods inside the stateful set.

The `-deployer` pipeline re-applies the stateful set. This will intentionally roll all the associated pods.

jonathanglassman replied:
Also, will the user know what "roll all the associated pods" means?

blairboy362 replied:
An SRE should definitely know what "roll the associated pods" means; we shouldn't be trying to turn this document into a "k8s fundamentals 101".

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,96 @@
+   through sandbox via the full resize process to make sure we haven't
+   broken anything before going to the Verify cluster.
+
+1. Pods in StatefulSets each have a volume. For Kubernetes to be able
```
is it StatefulSet, statefulset, or stateful set?

blairboy362 replied:
The k8s docs refer to StatefulSet pretty consistently, so we'll go with that!

jonathanglassman commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,96 @@
+   The `-deployer` pipeline will re-apply the statefulset. This will
+   intentionally roll all the associated pods.
+
+1. Once the pods are in a `Running` state, edit the
```
How do I know the pods are in a `Running` state?

blairboy362 replied:
Are you suggesting we include a command example here? As long as we don't have to explain all the options passed to it, shouldn't be too much of a digression.

jonathanglassman replied:
Depends on if we are confident the user knows how to check this? If you think we need to tell people this then I would include it e.g.
" Check that the pods are in a `Running`state: 
```
CODE
```"
However, if this isnt necessary, then you can keep it as is.

blairboy362 replied:
I think the target audience for this document will know how to check. So I think I'll leave that for now.

jonathanglassman reviewed (changes requested):
See comments and queries

philandstuff commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
````diff
@@ -0,0 +1,93 @@
+   $ gds sandbox kubectl -n gsp-system delete statefulset --cascade=false gsp-concourse-worker
+   ```
+
+   You should destroy the StatefulSet when the Concourse release pipeline starts to fail. You will see an error message similar to:
````
It feels to tell someone *when* to run a command after you've already told them to run it.  I can imagine going through this document, running the `delete statefulset` line, and then reading this line and thinking "wait why didn't you tell me before"

jonathanglassman replied:
Fair point. If you move this to before "To destroy the StatefulSet, run:..."  that should be ok hopefully. It pushes the action further away from the start of the numbered step, but you do need the context.

philandstuff commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,93 @@
+   intentionally roll all the associated pods.
+
+1. Once the pods are in a `Running` state, edit the
+   `PersisentVolumeClaim`s manually.
```
Is it worth saying *why* we have to do this?  If I've edited `persistence.worker.size` as described above, why do I also need to edit the PVCs?

blairboy362 replied:
I've added some context. Not sure how much detail we want to go into.

philandstuff replied:
@blairboy362 I can't see what you changed?

blairboy362 replied:
I find it's much easier for others to see the changes when they've actually been pushed.

philandstuff commented on docs/gds-supported-platform/resizing-persistent-volumes.md:
```diff
@@ -0,0 +1,93 @@
+      NAME                                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+      concourse-work-dir-gsp-concourse-worker-0   Bound    pvc-a633fb0a-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
+      concourse-work-dir-gsp-concourse-worker-1   Bound    pvc-a6366add-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
+      concourse-work-dir-gsp-concourse-worker-2   Bound    pvc-a6386683-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
```
above, the text says `Pods in StatefulSets each have a volume`, but here it seems like all pods share the same volume?

blairboy362 replied:
The volume IDs are different...

philandstuff replied:
oops, yes they are! 🙈

philandstuff committed:
Add context to resizing persistent volumes docs

I tried to split the context from the cookbook - if you just want to
follow instructions you can, but if you keep thinking "but why tho"
you can read about it at the bottom.
